frame/bridge/crab/issuing/src/lib.rs

```diff
@@ -42,73 +46,76 @@ mod types {
 	type RingCurrency<T> = <T as Config>::RingCurrency;
 }
 
-// --- substrate ---
-use frame_support::{
-	decl_error, decl_event, decl_module, decl_storage,
-	traits::{Currency, Get},
-};
-use sp_runtime::{traits::AccountIdConversion, ModuleId};
-// --- darwinia ---
-use types::*;
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::{Currency, Get};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::{traits::AccountIdConversion, ModuleId};
+	use types::*;
 
-pub trait Config: frame_system::Config {
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		#[pallet::constant]
+		type ModuleId: Get<ModuleId>;
 
-	type ModuleId: Get<ModuleId>;
+		type RingCurrency: Currency<AccountId<Self>>;
 
-	type RingCurrency: Currency<AccountId<Self>>;
+		type WeightInfo: WeightInfo;
 
-	type WeightInfo: WeightInfo;
-}
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+	}
 
-decl_event! {
-	pub enum Event<T>
-	where
-		AccountId = AccountId<T>,
-		RingBalance = RingBalance<T>,
-	{
+	// Define the pallet struct placeholder, various pallet function are implemented on it.
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::event]
+	pub enum Event<T: Config> {
 		/// Dummy Event. [who, swapped *CRING*, burned Mapped *RING*]
-		DummyEvent(AccountId, RingBalance, MappedRing),
+		DummyEvent(AccountId<T>, RingBalance<T>, MappedRing),
 	}
-}
 
-decl_error! {
-	pub enum Error for Module<T: Config> {
+	#[pallet::storage]
+	#[pallet::getter(fn total_mapped_ring)]
+	pub(super) type TotalMappedRing<T: Config> = StorageValue<_, MappedRing>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T> {
+		pub total_mapped_ring: MappedRing,
+		pub phantom: std::marker::PhantomData<T>,
```

Comment:
To be compatible with mock, added a phantom here, not sure it's a good option, any suggestions? @AurevoirXavier 
Because the following code builds storage with a type Test, so the GenesisConfig needs to be defined with a T which not used actually. To avoid a build error `unused generic T xxx`, the PhantomData was added.

```rust
        let mut t = frame_system::GenesisConfig::default()
                .build_storage::<Test>()
                .unwrap();
```

Reply:
We don't need the generic config.

Just keep it simple/

Reply:
I've fixed that.